server/backends/chunkstore/chunkstore.go

```diff
@@ -0,0 +1,413 @@
+	case io.SeekCurrent:
+		offset = r.off + offset
+	case io.SeekEnd:
+		for {
+			if exists, err := r.nextChunkExists(); err != nil {
+				return r.off, err
+			} else if !exists {
+				break
+			}
+			r.getNextChunk()
+		}
+		offset += r.off + int64(len(r.chunk)-r.chunkOff)
+	default:
+		return r.off, syscall.EINVAL
+	}
+
+	if offset < 0 {
+		return r.off, syscall.EINVAL
+	}
+
+	if offset >= r.off-int64(r.chunkOff) && offset < r.off+int64(len(r.chunk)-r.chunkOff) {
+		r.advanceOffset(offset - r.off)
+		return r.off, nil
+	}
+
+	if offset < r.off {
+		r.chunkIndex = r.startIndex
+		r.chunk = []byte{}
+		r.off = 0
+		r.chunkOff = 0
+	}
+	r.advanceOffset(offset - r.off)
+	return r.off, nil
+}
+
+func (r *chunkstoreReader) Close() error {
+	r.blobName = ""
+	r.chunk = nil
+	r.ctx = nil
+	r.chunkstore = nil
+	return nil
+}
+
+func (c *Chunkstore) Reader(ctx context.Context, blobName string) *chunkstoreReader {
+	return &chunkstoreReader{
+		chunkstore: c,
+		chunkIndex: math.MaxUint16,
+		startIndex: math.MaxUint16,
+		chunk:      []byte{},
+		blobName:   blobName,
+		ctx:        ctx,
+	}
+}
+
+func (c *Chunkstore) ReverseReader(ctx context.Context, blobName string) (*chunkstoreReader, error) {
+	chunkIndex := uint16(0)
+	for {
+		if exists, err := c.chunkExists(ctx, blobName, chunkIndex); err != nil {
+			return nil, err
+		} else if !exists {
+			break
+		}
+		chunkIndex++
+	}
+	return &chunkstoreReader{
+		chunkstore: c,
+		chunkIndex: chunkIndex,
+		startIndex: chunkIndex,
+		chunk:      []byte{},
+		blobName:   blobName,
+		ctx:        ctx,
+		reverse:    true,
+	}, nil
+}
+
+type writeResult struct {
+	err  error
+	size int
+}
+
+type chunkstoreWriter struct {
+	ctx                context.Context
+	chunkstore         *Chunkstore
+	writeChannel       chan []byte
+	writeResultChannel chan writeResult
+	blobName           string
+	blockSize          int
+	timeoutDuration    time.Duration
+}
+
+func (w *chunkstoreWriter) Write(p []byte) (int, error) {
+	w.writeChannel <- p
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return 0, os.ErrClosed
+		}
+		return result.size, result.err
+	case <-time.After(20 * time.Second):
+		return 0, context.DeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) Flush() (int, error) {
+	w.writeChannel <- nil
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return 0, os.ErrClosed
+		}
+		return result.size, result.err
+	case <-time.After(20 * time.Second):
+		return 0, context.DeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) Close() error {
+	close(w.writeChannel)
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return fmt.Errorf("Error closing %v: %w", w.blobName, os.ErrClosed)
+		}
+		return result.err
+	case <-time.After(w.timeoutDuration):
+		return os.ErrDeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) writeLoop() {
+	chunkIndex := uint16(0)
+	chunk := []byte{}
+	never := time.Unix(1<<63-62135596801, 999999999)
```

Comment:
could you make this a const? or maybe we can remove it entirely? Seems ok if a select case fires that checks for a flush but we don't actually flush anything because there's nothing ready.

Reply:
I can make the number a const, but not the `time.Time` (due to it not being a "character, string, boolean, or numeric value"), I tried :P Still, probably worth consting the number if nothing else. I was doing the thing you described before, where it would periodically flush nothing every two minutes, but we talked about it and you pointed out that since we never do anything when it wakes up that it shouldn't wake up, hence this solution. If you prefer the old method (maybe with a longer delay than two minutes, but not the maximum possible time representable in go?), I would be happy to do that instead.